tests/test_models.py

```diff
@@ -222,6 +223,18 @@ def test_get_stats(requests_mock):
     }
 
 
+def test_get_stats_when_token_expired(requests_mock):
+    # sadly, it just returns the global leaderboard, rather than a http 4xx
+    user = User("token12345678")
+    requests_mock.get(
+        url="https://adventofcode.com/2019/leaderboard/self",
+        text="<article><p>Below is the <em>Advent of Code 2021</em> overall leaderboard</p></article>"
+    )
+    expected_msg = "the auth token ...5678 is expired or not functioning"
+    with pytest.raises(DeadTokenError(expected_msg)):
```

Comment:
I see this is already merged, but you're using [`pytest.raises`](https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest.raises) incorrectly. It takes a `Type[Exception]`, not an `Exception`. You can pass an optional `match` keyword argument to check the message.

Reply:
Nope, this is correct. I don't like to use the `match` keyword, because I don't think it's obvious enough that the matching is done with a regex pattern rather than an exact string match. You can easily forget to put the `^anchors$` and re escaping on your string if you aren't aware the match is actually an re pattern.

Reply:
It's not correct. There is no form of `pytest.raises` that takes an `Exception` instance as an argument.

The docs (linked in my prior comment) says:

> expected_exception ([Type](https://docs.python.org/3/library/typing.html#typing.Type)[E] | [Tuple](https://docs.python.org/3/library/typing.html#typing.Tuple)[[Type](https://docs.python.org/3/library/typing.html#typing.Type)[E], ...]) – The excpected exception type, or a tuple if one of multiple possible exception types are excepted.

(yes, the "excpected" typo is in the docs 😆 I opened pytest-dev/pytest#10449 to fix it while writing this comment.)

You can see the [source](https://docs.pytest.org/en/7.2.x/_modules/_pytest/python_api.html#raises).

Reply:
Sorry, it's correct. If it were incorrect, this test (and similar ones) would fail with a `TypeError`. The tests are passing.

Reply:
I created an example project to demonstrate why I believe the test(s) in this project are wrong. Frankly, I'm unclear why we're seeing a difference in behavior.

See the relevant [source code](https://github.com/cj81499/pytest-example/blob/main/src/pytest_example/__init__.py) and [test cases](https://github.com/cj81499/pytest-example/blob/main/tests/test_example.py). See also that both [`mypy`](https://github.com/cj81499/pytest-example/actions/runs/3358718498/jobs/5566025971#step:7:15) and [`pytest`](https://github.com/cj81499/pytest-example/actions/runs/3358718498/jobs/5566025971#step:8:28) failed in ci.